### ISSUE TITLE
[Bug] in v-add-fastcgi-cache duration will not work

### ISSUE THREAD

**Issue by erikdemarco** [hestiacp/hestiacp#2792]
### Describe the bug

The way v-add-fastcgi-cache duration work is by adding the user setting value to this line:
https://github.com/hestiacp/hestiacp/blob/d50f95cf208049dfb6ac67a8020802121745bd60/bin/v-add-fastcgi-cache#L71

But, If you see this line:
https://github.com/hestiacp/hestiacp/blob/d50f95cf208049dfb6ac67a8020802121745bd60/bin/v-add-fastcgi-cache#L84

There is setting for inactive:30m it means if the cache is not requested again for 30 minutes, it will get pruned. see:
http://nginx.org/en/docs/http/ngx_http_proxy_module.html
"Cached data that are not accessed during the time specified by the inactive parameter get removed from the cache regardless of their freshness."

So, let say user enable fcgi cache for duration of 1day :
`v-add-fastcgi-cache user domain.tld 1d`

Its doesnt work, because it will get pruned if the url last accessed is more than 30minutes

### Tell us how to replicate the bug

1.  enable the domain to be cached for 1 day
`v-add-fastcgi-cache user domain.tld 1d`

2. create php file with this header. this to tell nginx to cache it for 1day
```
header("Cache-Control: s-maxage=86400, public");
header("X-Accel-Expires: 86400");
```

3. Test, and see the header sent in browser, is will be HIT after 2nd try. and test it again after 30minutes passed. it will be MISS

### Which components are affected by this bug?

(Backend) Web Server (Nginx, Apache2)

### Hestia Control Panel Version

1.6

### Operating system

ubuntu 20

**Comment by jaapmarcus** [comment]
So incases where duration is set to > 30m the  inactive time need to be increase to at least > $duration

**Comment by erikdemarco** [comment]
Honestly i a bit not agree to change default nginx's fastcgi_cache_valid. 
This setting tells nginx how long should we cache the url if there is no cache header (ex. cache-control) in the response. if we want to set proxy_cache_valid its better to set it low, because this setting takes very low priority compared to cache-control from header.

So if there is no cache header and user is just debugging the php code, they will be confuse why the code not changing (because of fastcgi_cache_valid takes place). 

maybe we should only change the inactive setting? This will act like a "max cache duration". So the user/app will decide it themself via cache header. Example

We set the v-add-fastcgi-cache to 1 day:
-) The app add cache-control for 1 hour. Nginx will only cache this for 1 hour
-) The app add cache-control for 7 day. Nginx will only cache this for 1 day (max cache duration we set)

And if you see this line:
https://github.com/hestiacp/hestiacp/blob/d50f95cf208049dfb6ac67a8020802121745bd60/bin/v-add-fastcgi-cache#L73

404 is cached to long, usually it cached for just 1minute

I think the most fit setting for proxy_cache_valid is we hardcoded it like this:
proxy_cache_valid 200 302 10m;
proxy_cache_valid 301      1h;
proxy_cache_valid any      1m;

Then we just play with the inactive setting?

**Comment by ioannidesalex** [comment]
@erikdemarco you are right that the inactive value will delete pages not accessed in the last 30 minutes, regardless of $duration. That's a very interesting point and we gave it a lot of thought to decide on those values.

However, every usecase is  different and we need to find some out of the box defaults that make sense for the majority of the users.

Here are my thoughts:

Let's take the case where a user is hosting 300 websites on his server. If we put this limitless, or 24h, it's going to need a very big storage space to keep all these files.

Given that fastcgi_cache_path is a global variable it's not possible to set inactive period = $duration for each site. Unless we are creating a new zone for each website. But still it doesn't mean that each user wants the inactive period to = $duration. And the reason is that Nginx has the capability (with serve stalled content option) to serve files in its cache in cases PHP/MySQL are down due to issues. So if a user sets $duration to 1m, it doesn't mean that he wants to delete all cache files in 1m, because otherwise he is not going to be able to serve them in a case of PHP/MySQL issue.

Nginx cache even though it works really well, is much more limited in terms of customization compared to other solutions such as Varnish. It was primarily designed as a "micro cache", used to cache for a few seconds or minutes. If I cache item it's not accessed within 30 minutes, then caching doesn't really help anyway, since there would be less than 48 requests per day.

To sum up, I believe that the values we have set as defaults cover the majority of cases. We can't really find a solution that covers every use case and keeps everyone satisfied and especially safe out of the box.

Also, please note that fastcgi_cache and proxy_cache are two different things, even though they work the same way. Proxy cache is for the cases where Nginx acts as a Reverse Proxy, where Fastcgi cache is for cases where Nginx is the Primary web server.

**Comment by erikdemarco** [comment]
@ioannidesalex 

"Let's take the case where a user is hosting 300 websites on his server. If we put this limitless, or 24h, it's going to need a very big storage space to keep all these files."

This not going to happen, because if you see on this line, we already set the max_size of cache file for each domain to 512mb:
https://github.com/hestiacp/hestiacp/blob/d50f95cf208049dfb6ac67a8020802121745bd60/bin/v-add-fastcgi-cache#L84

"So if a user sets $duration to 1m, it doesn't mean that he wants to delete all cache files in 1m, because otherwise he is not going to be able to serve them in a case of PHP/MySQL issue."
Lets see from the other point of view. Let say user is set the duration/inactive to default 30minutes. And this user install cache plugin (most wp user use cache plugin). this cache plugin uaully will cache the page for couple hours before the cache is expire via cache-control header. Because of our nginx setting this cache will be deleted in 30minutes which wasting our resources. This is for 1 page. How if user have many pages? 30 minutes expiration will seems like we dont use cache at all because for site with many pages, user randomly load the page.

Setting inactive to 24 hours doesnt mean the page will be cached for 24 hours. If the the plugin/script set the cache header to lower than 24 hour, it will respect that. The only problem if the script set cache header higher than "inactive". It will not respect that. Because inactive is the higher priority in this case.
Nginx is smart, if the max_size is reached. it will selectively prune which pages is rarely used and delete its cache. So if we use to low "inactive" this feature will not be used efficiently.

"Nginx cache even though it works really well, is much more limited in terms of customization compared to other solutions such as Varnish."
I 100% agree on this, even nginx cant do simple 'OR' comparison, but this simplicity and limited feature makes it very fast. faster than varnish. And some user doesnt want to add more difficulty in technology stack.

**Comment by ioannidesalex** [comment]
Still, 300x512GB =150GB of space, which users might not have available. Or what happens if for example he sets /var/cache/nginx as a tmpfs in order to put those cache files in memory to make it even faster? Also, what about a HestiaCP instance serving 2-3 huge sites that 512MB are not enough to keep the cache?

Also, using a cache plugin with FastCGI cache enable doesn't make a lot of sense and its counter productive in most cases.

Don't get me wrong, your points make a lot of sense and I completely understand your point of view. In fact, they are closer to my own, personal configuration/liking. What I am trying to say is that there are no defaults that can cover each and every case. We can keep debating for example cases for hours without finding a solution that covers everything.

HestiaCP is a panel bult for sysadmins. It provides a foundation for easy management and we aim to provide as sensible as possible defaults, with security and performance in mind. But configuration is needed, according to each system's needs.

I believe that the current defaults are pretty good for the vast majority of cases. We didn't receive any complains about overloading system etc. But that's my own opinion, maybe some other team members can also comment on this.

**Comment by erikdemarco** [comment]
"Still, 300x512GB =150GB of space, which users might not have available. " 
Then why we need to set the default max_cache to 512 mb in the first place? If we not going to use that huge, we should lower it for most of the user. if we really after most of the user.

"Also, using a cache plugin with FastCGI cache enable doesn't make a lot of sense and its counter productive in most cases."
Even not using a plugin, fastcgi will still cache evrything which has "cache-control" header. For some of my site which have more traffic i dont use such plugin. I depend only on nginx cache via "cache-control" header

**Comment by ioannidesalex** [comment]
Because we are using a single zone, where the combintation of 512 MB and 30 minutes of inactivity makes it highly unlikely to be reached or surpassed. Defining a MB limit is a safety net, e.g. in case there is an attack using query strings that generate extra cache copies.

Anyway, that's my personal opinion, I might be wrong though and I just see things differently than the rest.

Let's wait for some feedback from team members.

@jaapmarcus @ScIT-Raphael  what do you think?

**Comment by erikdemarco** [comment]
"Because we are using a single zone, where the combintation of 512 MB and 30 minutes of inactivity makes it highly unlikely to be reached or surpassed."

If we see this line, from my understanding $domain is for each domain, thus this 512m setting is for each domain. not a combination. correct me if im wrong:
https://github.com/hestiacp/hestiacp/blob/d50f95cf208049dfb6ac67a8020802121745bd60/bin/v-add-fastcgi-cache#L84

"Defining a MB limit is a safety net, e.g. in case there is an attack using query strings that generate extra cache copies."
This is not the good way to protect it. we have 'limit_req' to protect for such kind of attack

**Comment by jaapmarcus** [comment]
Default settings here are fine for the most use cases.  

It can be easily changed by extending the form it self and only allow admins to change the values. 

https://github.com/hestiacp/hestiacp/blob/5ed350f840ad84cb9793b5afd5b49befafb8bd8f/bin/v-add-fastcgi-cache#L69-L85

Question do we want to invest the extra time for such small user base who want it more flexible.  I say currently not there are other "problems" that should have a higher priority. 

@erikdemarco  feel free to submit a PR for changing it.

**Comment by ioannidesalex** [comment]
You are right for the first part, sorry about that, I made a mistake. What I wanted to say is that we are using a single zone per site.

Regarding throttling with limit_req: It doesn’t safeguard you with regards to cache size. Even a limit of 5r/s is 432k requests per day.

What would be ideal but takes some time to implement is to make all values user configurable via the panel. This way, anyone can chose what values he/she sees fit.

**Comment by ScIT-Raphael** [comment]
> I say currently not there are other "problems" that should have a higher priority.

Same opinion here, as it has been deeply discussed and explained, I'll close the bug report.
